pkg/secrethub/http/http.go

```diff
@@ -87,43 +89,44 @@ type ClientOptions struct {
 	Timeout   time.Duration
 }
 
-// httpClient is a raw client for the SecretHub http API.
-type httpClient struct {
-	client        *http.Client
-	authenticator auth.Authenticator
-	base          string // base url
-	version       string
+type AuthProvider interface {
+	Provide(httpClient *Client) (auth.Authenticator, error)
 }
 
-// newHTTPClient configures a new httpClient and overrides default values
+// Client is a raw client for the SecretHub http API.
+type Client struct {
+	client       *http.Client
+	authProvider AuthProvider
+	base         string // base url
+	version      string
+}
+
+// NewClient configures a new Client and overrides default values
 // when opts is not nil.
-func newHTTPClient(authenticator auth.Authenticator, opts *ClientOptions) *httpClient {
+func NewClient(with ...Option) *Client {
 	serverURL := DefaultServerURL
 	timeout := DefaultTimeout
-	if opts != nil {
-		if opts.ServerURL != "" {
-			serverURL = opts.ServerURL
-		}
-
-		if opts.Timeout > 0 {
-			timeout = opts.Timeout
-		}
-	}
-
 	serverURL = strings.TrimSuffix(serverURL, "/")
 	serverURL = serverURL + baseURLPath
 
-	return &httpClient{
+	client := &Client{
 		client: &http.Client{
 			Timeout: timeout,
 		},
-		authenticator: authenticator,
-		base:          serverURL,
-		version:       ClientVersion,
+		base: serverURL,
+		//version: secrethub.ClientVersion,
 	}
+	client.Options(with...)
+	return client
 }
 
-func (c *httpClient) Authenticate(in interface{}) (*api.Session, error) {
+func (c *Client) Options(with ...Option) {
```

Comment:
See also this comment (https://github.com/secrethub/secrethub-go/pull/95#issuecomment-520384922) about what needs to be done if we want to support changing options on existing clients.

Reply:
The reason is that it makes the usage a lot more straightforward. Otherwise we first have to gather all options that concern the HTTP Client before constructing the HTTP Client. With this function we can just create an HTTP Client and then add the options incrementally. In practice this is only done on init (because only the `secrethub` package has access to the HTTP Client and only calls `Options()` on init.) So this should not cause any problems. We could try to rewrite it, but I think it would become overly complex for something that is not exposed to the user (`http` is in `internals/`).

`With` and `WithOptions` indicate that it _returns_ a new client (at least to me). `ApplyOptions` okay?